budgetportal/models.py

```diff
@@ -1412,6 +1416,148 @@ class PackageWithoutGroupException(Exception):
+
+        params = {'sql': sql}
+        projects_result = requests.get(CKAN_DATASTORE_URL, params=params)
+        projects_result.raise_for_status()
+        revenue_data = projects_result.json()['result']['records']
+
+        # Assume project names are unique within the subset of featured projects
+        unique_project_names = []
+        for obj in revenue_data:
+            if obj['Project name'] not in unique_project_names:
+                unique_project_names.append(obj['Project name'])
+
+        projects = []
+        for project_name in unique_project_names:
+            project_list = filter(lambda x: x['Project name'] == project_name, revenue_data)
+            projects.append(InfrastructureProject(records=project_list))
+        return projects
+
+    def get_url_path(self):
+        return "/infrastructure-projects/{}".format(slugify(self.department_name + '-' + self.name))
+
+    def get_projected_expenditure(self):
+        # Include only projected amounts
+        projected_project_list = filter(lambda x: int(x['Financial Year']) > 2018, self.records)
+
+        projected_expenditure = 0
+        for project in projected_project_list:
+            projected_expenditure += float(project['Amount'])
+        return projected_expenditure
+
+    def get_cleaned_coordinates(self):
+        try:
+            if 'and' in self.gps_codes:
+                gps_codes_grouped = self.gps_codes.split('and')
+                for code_group in gps_codes_grouped:
+                    lat_long = [float(x) for x in code_group.split(',')]
+                    self.coordinates.append({
+                        'latitude': lat_long[0],
+                        'longitude': lat_long[1]
+                    })
+            elif ',' in self.gps_codes:
+                lat_long = [float(x) for x in self.gps_codes.split(',')]
+                self.coordinates.append({
+                    'latitude': lat_long[0],
+                    'longitude': lat_long[1]
+                })
+            else:
+                logger.warning("Invalid co-ordinates for infrastructure project '{}': {}".
+                               format(self.name, self.gps_codes))
+        except Exception as e:
+            logger.warning("Caught Exception '{}' for co-ordinates for infrastructure project '{}'".format(
+                e, self.name
+            ))
+        return self.coordinates
+
+    def get_provinces(self):
+        params = {'type': 'PR'}
+        if not self.coordinates:
+            self.get_cleaned_coordinates()
+        for c in self.coordinates:
+            province_result = requests.get(MAPIT_POINT_API_URL.format(c['longitude'], c['latitude']), params=params)
+            province_result.raise_for_status()
+            list_of_objects_returned = province_result.json().values()
+            if len(list_of_objects_returned) > 0:
+                province_name = list_of_objects_returned[0]['name']
+                if province_name not in self.provinces:
```

Comment:
you could use a [set](https://docs.python.org/2/library/sets.html) as an intermediate datastructure, and just turn it into a list when this method is done so you don't need to do `if province_name not in ...`